File: baekJoon/baekjoon_14923.cpp
// 미로 탈출
// 4:05

#include <iostream>
#include <queue>

using namespace std;

int n, m;
bool arr[1001][1001];
bool visited[1001][1001][2] = { { {false, } ,}, };
int dx[4] = { 1, -1, 0, 0 };
int dy[4] = { 0, 0, 1, -1 };

int bfs(int sy, int sx, int ly, int lx) {
	queue<pair<bool, pair<int, int>>> q;
	q.push(make_pair(true, make_pair(sy, sx)));
	visited[sy][sx][true] = true;
	int y, x, ny, nx, len, t = 0;
	bool flag;
	while (!q.empty()) {
		len = q.size();
		while (len--) {
			y = q.front().second.first;
			x = q.front().second.second;
			flag = q.front().first;
			q.pop();
			if (y == ly && x == lx) return t;

			for (int i = 0; i < 4; i++) {
				ny = y + dy[i];
				nx = x + dx[i];
				if (ny < 0 || ny == n || nx < 0 || nx == m) continue;
				if (arr[ny][nx]) {
					if (flag) {
						if (!visited[ny][nx][false]) {
							visited[ny][nx][false] = true;
							q.push(make_pair(false, make_pair(ny, nx)));
						}
					}
				}
				else {
					if (!visited[ny][nx][flag]) {
						visited[ny][nx][flag] = true;
						q.push(make_pair(flag, make_pair(ny, nx)));
					}
				}
			}
		}
		t++;
	}
	return -1;
}

int main() {
	cin.sync_with_stdio(0);
	cin.tie(0);
	int sy, sx, dy, dx;
	cin >> n >> m >> sy >> sx >> dy >> dx;
	for (int i = 0; i < n; i++)
		for (int j = 0; j < m; j++)
			cin >> arr[i][j];
	cout << bfs(sy - 1, sx - 1, dy - 1, dx - 1) << '\n';
	return 0;
}

File: baekJoon/baekjoon_14923.py
# 미로 탈출

import sys
from collections import deque
input = sys.stdin.readline

def bfs(n, m, y, x, fy, fx, arr, visited):
    dq = deque()
    dq.append([y, x, 0])
    visited[y][x][0] = True
    dy = [0, 0, 1, -1]
    dx = [1, -1, 0, 0]

    t = 0
    while dq:
        t += 1
        sz = len(dq)
        while sz > 0:
            y, x, flag = dq.popleft()
            sz -= 1
            for i in range(4):
                ny = y + dy[i]
                nx = x + dx[i]
                if ny == fy and nx == fx: return t
                if ny < 0 or ny == n or nx < 0 or nx == m or visited[ny][nx][flag]: continue
                if flag == 0:
                    if arr[ny][nx] == 0:
                        dq.append([ny, nx, flag])
                        visited[ny][nx][flag] = True
                    else:
                        if visited[ny][nx][1]: continue
                        dq.append([ny, nx, 1])
                        visited[ny][nx][1] = True
                else:
                    if arr[ny][nx] == 0:
                        dq.append([ny, nx, flag])
                        visited[ny][nx][flag] = True
                    else:
                        continue

    return -1

n, m = map(int, input().split())
y, x = map(int, input().split())
fy, fx = map(int, input().split())
arr = [list(map(int, input().split())) for _ in range(n)]
visited = [[[False for _ in range(2)] for _ in range(m)] for _ in range(n)]
print(bfs(n, m, y-1, x-1, fy-1, fx-1, arr, visited))
